pkg/framework/controller_watcher.go

```diff
@@ -139,3 +132,20 @@ func (c *controllerWatcher) Watch(watchedObjects ...WatchedObjects) (err error)
 
 	return
 }
+
+func (c *controllerWatcher) getServerGroupMap() (map[string]map[string]bool, error) {
+	serverGroups, err := c.discoverClient.ServerGroups()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't fetch server groups list: %v", err)
+	}
+
+	serverGroupMap := map[string]map[string]bool{}
+	for _, serverGroup := range serverGroups.Groups {
+		versions := make(map[string]bool)
+		for _, version := range serverGroup.Versions {
```

Comment:
Instead of two maps, I think we could have only one with the `object.GroupVersion.String()` instead. wdyt? This way we'd decreased the complexity level as well. We could concatenate both of them with `fmt` like `version\group`. WDYT?

Reply:
I considered that when implementing it, but I was concerned with groups where the version wasn't necessarily at the end (`foo.bar/v1/foo`) which could make matching more difficult. We can definitely index the map with the help of`fmt`, good idea:

```
func (c *controllerWatcher) getServerGroupMap() (map[string]bool, error) {
	serverGroups, err := c.discoverClient.ServerGroups()
	if err != nil {
		return nil, fmt.Errorf("couldn't fetch server groups list: %v", err)
	}

	serverGroupMap := make(map[string]bool)
	for _, serverGroup := range serverGroups.Groups {
		for _, version := range serverGroup.Versions {
			key := fmt.Sprintf("%s/%s", serverGroup.Name, version.Version)
			serverGroupMap[key] = true
		}
	}
	return serverGroupMap, nil
}
```

The downside I see is that it would make access to the map more complex, as we'd need to build the key when accessing the map as well:

```
for _, object := range watchedObjects {
		// core resources
		if object.AddToScheme == nil {
			desiredObjects = append(desiredObjects, object)
		} else {
			key := fmt.Sprintf("%s/%s", object.GroupVersion.Group, object.GroupVersion.Version)
			if _, found := serverGroupMap[key]; found {
```

We could keep a const for format:

`const keyFormat = "%s/%s"`

What do you think? Indexing as `map[group/version]` or `map[group][version]`?

Reply:
When accessing the map, can't you just use `object.GroupVersion.String()` as key? That's the motivation behind creating the map like this. wdyt?

Reply:
Oh, yeah! I missed that, thanks :grin: